### PR TITLE
Added validation for annotations

### LINKED ISSUE #1992: Add RFC952 validation to Annotations
Similar to #1297, validate Annotations with the same modified RFC952 standard that is used for labels.

Include tests containing valid and invalid annotations to show their usage and update the documentation accordingly.

### PR DESCRIPTION
**Pull request by markturansky**
Annotations are validated use the same rules as Labels.  Resolves #1992.

### CODE REVIEW

**Comment by markturansky** [comment]
Please let me know what other types require annotation validation.  I pretty much copied the validation for labels.

**Comment by markturansky** [comment]
@bgrant0607

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -429,6 +430,17 @@ func ValidateLabels(labels map[string]string, field string) errs.ValidationError
 	return allErrs
 }
 
+// ValidateAnnotations validates that a set of annotations are correctly defined.
+func ValidateAnnotations(annotations map[string]string, field string) errs.ValidationErrorList {
```
Comment (thockin) [inline pkg/api/validation/validation.go]: This appears to be line-for-line identical to ValidateLabels() - why not just call that?  annotations is a label set...

**Comment by thockin** [comment]
Worth doing before Clayton's #3789 generification lands?

File: pkg/api/validation/validation.go (hunk 1 above)
Comment (markturansky) [inline pkg/api/validation/validation.go]: I cringed at the duplication, too.  I removed it and reused the ValidationLabels func.

**Comment by markturansky** [comment]
The tests are valid before and after #3789.  The changes to validation.go are pretty trivial for the generification work.

**Comment by thockin** [comment]
LGTM.  Cayton can absorb this merge.

**Comment by derekwaynecarr** [comment]
@markturansky - can you verify this will not break OpenShift annotations on a deploy object?

**Comment by bgrant0607** [comment]
Thanks much. Sorry for being slow.

**Comment by smarterclayton** [comment]
We use mixed case and periods, so we're broken.  Is mixed case an issue in annotation names?  I don't know that the arguments for making them the same ever depended on anything except "maybe we want to conflate fields and labels in the client api".  An example of an annotation with mixed case would be field names preserved during conversion for a non round tripped value.

**Comment by derekwaynecarr** [comment]
Since we don't plan to support query on an annotation, I don't get why they share the same validation rules as labels which can make some sensible restrictions on operator characters for label keys.  For me, an Annotation should really be opaque to kubernetes.

**Comment by bgrant0607** [comment]
Sorry. We'd always planned to make annotations consistent with labels. We're trying to avoid having lots of different rules for users to remember. I'd be ok with relaxing all of label, annotation, and resource keys, but I don't see a strong case for inconsistency.

I agree that annotation values should be opaque.

**Comment by smarterclayton** [comment]
It seems like the two use cases - system attached metadata, and query set syntax, really are different.  Are they even used by the same class of users?  I don't mind labels being restrictive to aid in approachability and queryability, but I can imagine a wide range of possible keys for annotations matching with automated processes that need to adapt values from external systems.  But that's just an opinion - convenience for one class of use can be inconvenience for the next class.

**Comment by thockin** [comment]
I don't think relaxing label keys is particularly useful, though its
probably not harmful within bounds.

I can buy loosening annotations.  I did not give it much thought, frankly,
and I guess I agree they are essentially opaque.

**Comment by bgrant0607** [comment]
Are we just talking about case sensitivity and more dots? Allowing more characters, such as '=' and '/' would be problematic.

**Comment by smarterclayton** [comment]
Right now / is allowed in labels and annotations.  I don't know if annotations need spaces, although I'm sure someone at some point will ask for them.

**Comment by derekwaynecarr** [comment]
For me it's just a matter of how opaque is opaque, but I can defer to @smarterclayton on how much he wants to relax.  In practice, I think I should be able to stick base64 encoded annotation values.

**Comment by thockin** [comment]
If we want to call annotations opaque, then let's just call them opaque.

**Comment by thockin** [comment]
VALUES are free form in both labels and annotations.  I thought we were
debating keys...

**Comment by derekwaynecarr** [comment]
I also could see people wanting to stick some Xml snippet even if it's a bad idea for other reasons.  Agree with Tim that opaque should be opaque.

**Comment by derekwaynecarr** [comment]
Sorry, thanks for reminder Tim, then I am corrected.  I don't care as much on keys.  Been working too much :-)

**Comment by bgrant0607** [comment]
Yes, I explicitly want to permit json, xml, whatever in the values.

**Comment by smarterclayton** [comment]
Opened #4017 to slightly relax this until we close the discussion.  It did indeed break OpenShift, and if we decide to keep it strict it's going to take us a while to migrate.

**Pull request merged**